File: LRUCache.java
import java.util.HashMap;

import javax.lang.model.element.Element;

class Node {
  int value;
  Node next;
  Node prev;
  int key;

  public Node(int value, Node next, Node prev, int key) {
    this.value = value;
    this.key = key;
    this.next = next;
    this.prev = prev;
  }

  public Node(int key, int value) {

    this.value = value;
    this.key = key;
    this.next = null;
    this.prev = null;
  }

  public String toString() {
    return "(" + this.key + "->" + this.value + ")";
  }
}

public class LRUCache {

  int capacity;
  Node beginNode;
  Node endNode;
  int currentLength;
  HashMap<Integer, Node> map;

  public LRUCache(int capacity) {
    this.capacity = capacity;
    this.beginNode = null;
    this.endNode = null;
    this.currentLength = 0;
    this.map = new HashMap<>();
  }

  public int get(int key) {
    if (!map.containsKey(key)) {
      return -1;
    } else {
      var node = map.get(key);
      if (node.next != null && node.prev == null) {
        beginNode.next.prev = null;
        this.beginNode.prev = this.endNode;
        this.endNode.next = this.beginNode;
        this.endNode = this.beginNode;
        this.beginNode = this.endNode.next;
        this.endNode.next = null;
      } else if (node.next != null && node.prev != null) {
        node.next.prev = node.prev;
        node.prev.next = node.next;
        node.next = null;
        this.endNode.next = node;
        node.prev = this.endNode;
        this.endNode = node;
      }
      return node.value;
    }

  }

  public void put(int key, int value) {
    if (map.containsKey(key)) {
      this.get(key);
      this.map.get(key).value = value;
      return;
    }
    Node node = new Node(key, value);
    if (currentLength == 0) {
      beginNode = node;
      endNode = node;
      map.put(key, node);
      currentLength++;
      return;
    }
    if (capacity > currentLength) {
      this.endNode.next = node;
      node.prev = endNode;
      this.endNode = node;
      this.map.put(key, node);
      currentLength++;
      return;
    }
    this.endNode.next = node;
    node.prev = this.endNode;
    endNode = node;
    this.map.remove(beginNode.key);
    this.beginNode.next.prev = null;
    this.beginNode = this.beginNode.next;
    map.put(key, node);
  }
}

File: js/LRUCache.js
class Node{
  constructor(key,value) {
    this.value = value;
    this.key = key; 
    this.prev = null;
    this.next = null;
  }
}
class LRUCache{
  constructor(capacity){
      this.map = new Map();
      this.end = null;
      this.begin =null;
      this.capacity =capacity;
      this.currentLength = 0;
  }
  put(key, value){
    if(this.map.has(key)){
      this.get(key);
      this.map.get(key).value = value;
    }
    else{
      let node = new Node(key, value)
      if(this.currentLength ==0){
        this.begin = node;
        this.end = node;
      }
      else if(this.currentLength< this.capacity){
        this.end.next = node;
        node.prev = this.end;
        this.end = node;
      }
      else if( this.currentLength >= this.capacity){
        this.end.next = node;
        node.prev = this.end;
        this.end = node;
        this.begin.next.prev= null;
        this.map.delete(this.begin.key)
        this.begin = this.begin.next;
      }
      this.currentLength++;
      this.map.set(key, node)
    }
  }
  get(key){
    if(!this.map.has(key)){
      return -1;
    }
    else{
      let node =  this.map.get(key);
      if(node.next!= null && node.prev == null){
        node.next.prev = null;
        this.begin = node.next;
        node.next = null;
        this.end.next = node;
        node.prev = this.end;
        this.end = this.end.next;
      }
      else if(node.next!= null && node.prev != null){
        node.next.prev = node.prev;
        node.prev.next = node.next;
        node.prev = this.end;
        node.next = null;
        this.end.next = node;
        this.end = this.end.next;
      }
      return this.map.get(key).value;
    }
  }
}